File: 120-triangle/triangle.py
class Solution(object):
    def minimumTotal(self, triangle):
        """
        :type triangle: List[List[int]]
        :rtype: int
        """
        rows = len(triangle)

        dp = [[0] * rows for _ in range(rows)]
        
        dp[rows - 1] = triangle[rows - 1]

        for i in range(rows - 2, -1, -1):
            for j in range(i + 1):
                dp[i][j] = triangle[i][j] + min(dp[i + 1][j], dp[i + 1][j + 1])

        return dp[0][0]

File: 120-triangle/triangle.java
class Solution {
    public int minimumTotal(List<List<Integer>> triangle) {
        int rows = triangle.size();

        int[][] dp = new int[rows][rows];

        //row 끝에 값 채우기
        for (int i = 0; i < rows; i++) {
            dp[rows - 1][i] = triangle.get(rows - 1).get(i);
        }

        //DP로 값 업데이트
        for (int i = rows - 2; i >= 0; i--) {
            for (int j = 0; j <= i; j++) {
                dp[i][j] = triangle.get(i).get(j) + Math.min(dp[i + 1][j], dp[i + 1][j + 1]);
            }
        }

        return dp[0][0];
    }
}